jerry-core/lit/lit-strings.cpp

```diff
@@ -597,28 +597,41 @@ lit_read_code_point_from_utf8 (const lit_utf8_byte_t *buf_p, /**< buffer with ch
   return bytes_count;
 } /* lit_read_code_point_from_utf8 */
 
-
 /**
- * Calculate hash from last LIT_STRING_HASH_LAST_BYTES_COUNT characters from the buffer.
+ * Calc hash using the specified hash_basis.
  *
  * @return ecma-string's hash
  */
-lit_string_hash_t
-lit_utf8_string_calc_hash_last_bytes (const lit_utf8_byte_t *utf8_buf_p, /**< characters buffer */
-                                      lit_utf8_size_t utf8_buf_size) /**< number of characters in the buffer */
+lit_string_hash_t __attr_always_inline___
+lit_utf8_string_hash_combine (lit_string_hash_t hash_basis, /**< hash to be combined with */
+                              const lit_utf8_byte_t *utf8_buf_p, /**< characters buffer */
+                              lit_utf8_size_t utf8_buf_size) /**< number of characters in the buffer */
 {
   JERRY_ASSERT (utf8_buf_p != NULL || utf8_buf_size == 0);
 
-  lit_utf8_byte_t byte1 = (utf8_buf_size > 0) ? utf8_buf_p[utf8_buf_size - 1] : (lit_utf8_byte_t) 0;
-  lit_utf8_byte_t byte2 = (utf8_buf_size > 1) ? utf8_buf_p[utf8_buf_size - 2] : (lit_utf8_byte_t) 0;
+  uint32_t hash = hash_basis;
+
+  for (uint32_t i = 0; i < utf8_buf_size; i++)
+  {
+    hash = (hash ^ utf8_buf_p[i]) * 16777619;
```

Comment:
This is implementation of FNV hash function, which is released into public domain. Constants used, are carefully picked primes by the authors. More info: http://www.isthe.com/chongo/tech/comp/fnv/